blockchain/state/statedb.go

```diff
@@ -565,7 +565,7 @@ func (s *StateDB) SelfDestruct6780(addr common.Address) {
 		return
 	}
 
-	if stateObject.created {
+	if stateObject.created && stateObject.account.Type() == account.SmartContractAccountType {
```

Comment:
As of Cancun's EIP6780, I reckon that selfdestruct is only possible during the same transaction as create. Such usage is impossible in EOA with code because thebcode is created elsewhere before assigned tonthe EOA. So why is this condition necessary at all? Is there a related testcase?

Reply:
https://github.com/ethereum/execution-spec-tests/blob/pectra-devnet-4%40v1.0.1/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L416-L472
Yes. For example, this test failed if there wasn't the condition. Creating an EOA by setting a code and executing the EOA can be done in the same tx.

Reply:
I found a more fundamental reason. In geth EVM, there are three account creation functions.
- `statedb.CreateAccount()`
  - Create an empty stateObject, but does NOT set `obj.newContract`
  - Called by EVM.Call and EVM.create.
- `statedb.CreateContract()`
  - Set `obj.newContract = true`
  - Called by EVM.create.
- `obj.newContract` is equivalent to Kaia's `obj.created`. Compare the declarations in [geth](https://github.com/ethereum/go-ethereum/blob/f8f5609b8eced8421110a9193bbd6cefee40b4af/core/state/state_object.go#L79-L84) and [kaia](https://github.com/kaiachain/kaia/blob/dev/blockchain/state/state_object.go#L112-L113).
---
In kaia, there are four.
- `statedb.CreateAccount()`
  - Create an empty EOA and set `obj.created = true` inside `s.createObject()`
  - Only for testing
- `statedb.CreateEOA()`
  - Create an empty EOA and set `obj.created = true` inside `s.createObjectWithMap()`
  - Called by EVM.Call
- `statedb.CreateSmartContractAccount()`
  - Equivalent to CreateSmartContractAccountWithKey
  - Called by EVM.Call if addr is an precompile
- `statedb.CreateSmartContractAccountWithKey()`
  - Create an empty SCA and set `obj.created = true` inside `s.createObjectWithMap()`
  - Called by EVM.create

I propose to fix:
- `s.createObject()` and `s.createObjectWithMap()` does NOT set `obj.created`
- Only `statedb.CreateSmartContractAccountWithKey` sets `obj.created = true`
---
Your `type == SCA` fix was equivalent to geth's behavior because [obj was created during EVM.create] if and only if [obj was created and it's an SCA].

Reply:
I solved this at https://github.com/kaiachain/kaia/pull/149/commits/a98782125061836a6eae86f50abee7f8711d5dde.

Reply:
I checked the related code and making `created = true` only in `CreateSmartContractAccountWithKey` would work. 

I also found exceptional cases in which SCA didn't exist before `SetCode`: Injecting registry at Randao HF and multicall since Kaia HF. But those contracts are not affected by this change because it's guaranteed not to be destructed.